### PR TITLE
Flash support

### PR DESCRIPTION
**Pull request by andresv**
**WIP**: do not merge yet.

This PR adds support to unlock, erase page, write to and read from STM32 flash memory.

### CODE REVIEW

File: src/flash.rs [hunk 1]
```diff
@@ -0,0 +1,246 @@
+use crate::stm32::FLASH;
+
+/// The first address of flash memory
+pub const FLASH_START: u32 = 0x0800_0000;
+pub const FLASH_END: u32 = 0x0801_FFFF;
+
+/// The size of a Flash memory page, in bytes
+pub const PAGE_SIZE: u32 = 2048;
+/// How many Flash memory pages there are
+pub const NUM_PAGES: u32 = 64;
```
Comment (andresv) [inline src/flash.rs:10]: TODO: some devices have less memory.

File: src/flash.rs [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+    /// Returns a &[u8] if the address and length are valid.
+    /// Length must be a multiple of 4.
+    fn read(&mut self, address: u32, length: usize) -> Result<&'static [u8]> {
+        verify_address(address)?;
+        verify_length(length)?;
+        let address = address as *const _;
+        unsafe {
+            Ok(core::slice::from_raw_parts::<'static, u8>(address, length))
+        }
+    }
```
Comment (andresv) [inline src/flash.rs:117]: Is it actually sound?
```rust
let slice = f.read(addr, 8);
f.write(addr, data);
```
After that slice points to new data. Is it ok, or should we come up with somekind of reader writer locking mechanism.

Comment (andresv) [reply src/flash.rs:117]: It should be handled here for F1 flash with read/write grants: https://github.com/stm32-rs/stm32f1xx-hal/pull/141/files

File: src/flash.rs (hunk 1 above)
Comment (ijager) [reply src/flash.rs:11]: So we could have a feature to select the flash size, with largest option the default. And/Or have more specific device features (optionally). e.g. `--features stm32g071gb` 

Which are implemented as

```
stm32g071c8 = ["stm32g0/stm32g071", "stm32g0x1", "flash_64k"]
stm32g071gb = ["stm32g0/stm32g071", "stm32g0x1", "flash_128k"]
```

But we will need to define all possible options I guess

**Comment by andresv** [comment]
Reminder: flash size could be read out on runtime https://github.com/stm32-rs/stm32-device-signature/blob/master/src/lib.rs#L37-L41.

File: src/flash.rs (hunk 1 above)
Comment (andresv) [reply src/flash.rs:11]: I haven't yet looked how others have solved it.

**Comment by mettz** [comment]
Sorry, are there any updates to this?

**Comment by andresv** [comment]
This actually works, but it is not polished and it is possible to do stupid things.

**Comment by mettz** [comment]
I'm currently working on a project that requires flash, so if I can help advance this I will be glad to work on it

**Comment by andresv** [comment]
I think we should use [L4 flash traits](https://github.com/stm32-rs/stm32l4xx-hal/blob/master/src/traits/flash.rs) and driver itself should look like this: https://github.com/stm32-rs/stm32l4xx-hal/blob/master/src/flash.rs.

You could copy those flash traits here and modify my driver to use them.

**Comment by mettz** [comment]
Sounds good! Should I make another PR or add changes to this one? (I think the second option requires me to have write access to your fork, correct me if I am wrong)

**Comment by andresv** [comment]
Its ok if you just do a new one.

**Comment by andresv** [comment]
Flash support is added with this PR: https://github.com/stm32-rs/stm32g0xx-hal/pull/41

**Pull request closed without merging**